tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -2212,61 +2213,53 @@ llvm::Optional<Value> convertFusedActivation(PatternRewriter& rewriter,
 
   if (input_is_qtype) {
+    // We can always make output/input tensor's scale/zp always be the same when
+    // legalizing fused_activation_function, as it's generated during
+    // legalization.
     auto input_qtype =
         input_type.getElementType().cast<mlir::quant::UniformQuantizedType>();
 
-    if (fused_activation_fn.getValue() == "TANH") {
-      // TODO: implement with TABLE
-      op->emitWarning("Quantized TANH lowering TBD!");
-      return llvm::None;
-    } else {
-      RankedTensorType rescale_type = RankedTensorType::get(
-          input_type.getShape(), rewriter.getIntegerType(32));
-
-      Value op1_rescale_in = buildRescaleToInt32(
-          rewriter, op, input_value, 1.0f, input_qtype.getZeroPoint());
-
-      Value op2_relu_op1;
-      if (fused_activation_fn.getValue() == "NONE") {
-        return input_value;
-      } else if (fused_activation_fn.getValue() == "RELU") {
-        auto relu_op = rewriter.create<tosa::ReluNOp>(
-            op->getLoc(), rescale_type, op1_rescale_in,
-            rewriter.getI64IntegerAttr(std::numeric_limits<int32_t>::max()),
-            rewriter.getF32FloatAttr(0));
-
-        op2_relu_op1 = relu_op.getResult();
-
-      } else if (fused_activation_fn.getValue() == "RELU6") {
-        int64_t rescaled_6 = std::llround(6.0f / input_qtype.getScale()) +
-                             input_qtype.getZeroPoint();
-
-        auto relu_op = rewriter.create<tosa::ReluNOp>(
-            op->getLoc(), rescale_type, op1_rescale_in,
-            rewriter.getI64IntegerAttr(rescaled_6),
-            rewriter.getF32FloatAttr(0.0f));
-
-        op2_relu_op1 = relu_op.getResult();
-
-      } else if (fused_activation_fn.getValue() == "RELU_N1_TO_1") {
-        int64_t rescaled_n1 = std::llround(-1.0f / input_qtype.getScale()) +
-                              input_qtype.getZeroPoint();
-        int64_t rescaled_1 = std::llround(1.0f / input_qtype.getScale()) +
-                             input_qtype.getZeroPoint();
-
-        auto relu_op = rewriter.create<tosa::ClampOp>(
-            op->getLoc(), rescale_type, op1_rescale_in,
-            rewriter.getI64IntegerAttr(rescaled_n1),
-            rewriter.getI64IntegerAttr(rescaled_1),
-            rewriter.getF32FloatAttr(0.0f), rewriter.getF32FloatAttr(0.0f));
-
-        op2_relu_op1 = relu_op.getResult();
-      } else {
-        return llvm::None;
-      }
+    if (fused_activation_fn.getValue() == "NONE") {
+      return input_value;
+    } else if (fused_activation_fn.getValue() == "RELU") {
+      int32_t quantized_0 = input_qtype.getZeroPoint();
+      int32_t quantized_max = input_qtype.getStorageTypeMax();
+
+      auto clamp_op = rewriter.create<tosa::ClampOp>(
+          op->getLoc(), input_type, input_value,
+          rewriter.getI64IntegerAttr(quantized_0),
+          rewriter.getI64IntegerAttr(quantized_max),
+          rewriter.getF32FloatAttr(0), rewriter.getF32FloatAttr(0));
```

Comment:
per some other comments: what happens if `input_type` is not f32? (here and elsewhere). I feel you may be missing a helper that gets a type appropriate float attr given an input type.

Reply:
I agree we should do more check where we implicitly assume types are f32.

But isn't this function already did the check you're asking?
if (input_is_qtype) {
  ... // legalize for quantized type
} else {
   ... // legalize for f32
   // maybe I should check and return failure() if not input_type is f32 here
}